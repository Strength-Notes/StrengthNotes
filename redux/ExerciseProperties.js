const ExerciseProperties = {
  WEIGHT: 'Weight',
  REPS: 'Reps',
  TIME: 'Time',
  DISTANCE: 'Distance',
  NONE: 'None',
};

export default ExerciseProperties;
