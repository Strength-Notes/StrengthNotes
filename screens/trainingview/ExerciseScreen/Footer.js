import React from 'react';
import {
  Text,
  TouchableOpacity,
  TextInput,
  View,
  Picker,
} from 'react-native';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import ExerciseProperties from '../../../redux/ExerciseProperties';
import { addSetAction } from '../../../redux/actions';
import styles from './EntryTab.styles';

class Footer extends React.Component {
  constructor(props) {
    super(props);

    this.addSetDispatch = props.addSetDispatch;

    const { date, exercise } = props;

    this.state = {
      date,
      exercise,
    };
  }

  // Cannot be made into an arrow function, as it results in a bug with Expo
  // eslint-disable-next-line class-methods-use-this
  getExerciseInput({ property, state, setState }) {
    const {
      weightInput,
      repsInput,
      distanceInput,
      distanceUnitSelected,
      hoursInput,
      minutesInput,
      secondsInput,
    } = state;

    switch (property) {
      case ExerciseProperties.WEIGHT: {
        return (
          <TextInput
            style={styles.weightInput}
            onChangeText={(text) => {
              setState({ weightInput: text });
            }}
            value={weightInput}
            placeholder="Weight"
            keyboardType="decimal-pad"
            maxLength={7}
          />
        );
      }
      case ExerciseProperties.REPS: {
        return (
          <TextInput
            style={styles.repsInput}
            onChangeText={(text) => {
              setState({ repsInput: text });
            }}
            value={repsInput}
            placeholder="Reps"
            keyboardType="decimal-pad"
            maxLength={4}
          />
        );
      }
      case ExerciseProperties.DISTANCE: {
        return (
          <View style={[styles.compoundContainer, { flex: 2 }]}>
            <TextInput
              style={styles.distanceInput}
              onChangeText={(text) => {
                setState({ distanceInput: text });
              }}
              value={distanceInput}
              placeholder="Distance"
              keyboardType="decimal-pad"
              maxLength={7}
            />
            <Picker
              style={styles.distanceUnitInput}
              selectedValue={distanceUnitSelected}
              onValueChange={(selected) => {
                setState({ distanceUnitSelected: selected });
              }}
              mode="dropdown"
            >
              <Picker.Item label="m" value="m" />
              <Picker.Item label="km" value="km" />
              <Picker.Item label="yd" value="yd" />
              <Picker.Item label="mi" value="mi" />
            </Picker>
          </View>
        );
      }
      case ExerciseProperties.TIME: {
        return (
          <View style={[styles.compoundContainer, { flex: 3 }]}>
            <TextInput
              style={styles.timeInput}
              onChangeText={(text) => {
                setState({ hoursInput: text });
              }}
              value={hoursInput}
              placeholder="HH"
              keyboardType="decimal-pad"
              maxLength={2}
            />
            <TextInput
              style={styles.timeInput}
              onChangeText={(text) => {
                setState({ minutesInput: text });
              }}
              value={minutesInput}
              placeholder="MM"
              keyboardType="decimal-pad"
              maxLength={2}
            />
            <TextInput
              style={styles.timeInput}
              onChangeText={(text) => {
                setState({ secondsInput: text });
              }}
              value={secondsInput}
              placeholder="SS"
              keyboardType="decimal-pad"
              maxLength={5}
            />
          </View>
        );
      }
      default: {
        return <View />;
      }
    }
  }

  addSetHandler = () => {
    const { distanceUnitSelected } = this.state;
    let {
      weightInput,
      repsInput,
      distanceInput,
      hoursInput,
      minutesInput,
      secondsInput,
      rpeInput,
    } = this.state;

    weightInput = Number(weightInput);
    repsInput = Number(repsInput);
    distanceInput = Number(distanceInput);
    hoursInput = Number(hoursInput);
    minutesInput = Number(minutesInput);
    secondsInput = Number(secondsInput);
    rpeInput = Number(rpeInput);

    // Set all NaNs or undefs to 0, to prevent time becoming NaN
    if (!hoursInput) hoursInput = 0;
    if (!minutesInput) minutesInput = 0;
    if (!secondsInput) secondsInput = 0;

    const time = hoursInput * 3600 + minutesInput * 60 + secondsInput;

    this.addSetDispatch(
      this.state,
      weightInput,
      repsInput,
      distanceInput,
      distanceUnitSelected,
      time,
      rpeInput,
    );
  }

  render = () => {
    const { exercise } = this.state;
    const { primary, secondary } = exercise;

    return (
      <View style={styles.footerContainer}>
        <this.getExerciseInput
          property={primary}
          state={this.state}
          setState={(newState) => { this.setState(newState); }}
        />
        <this.getExerciseInput
          property={secondary}
          state={this.state}
          setState={(newState) => { this.setState(newState); }}
        />

        <TextInput
          style={styles.rpeInput}
          onChangeText={(text) => { this.setState({ rpeInput: text }); }}
          value={this.state.rpeInput} // eslint-disable-line
          placeholder="RPE"
          keyboardType="decimal-pad"
          maxLength={4}
        />
        <TouchableOpacity
          style={styles.addButton}
          onPress={this.addSetHandler}
        >
          <Text style={styles.addButtonText}>Add</Text>
        </TouchableOpacity>
      </View>
    );
  };
}

Footer.propTypes = {
  date: PropTypes.string.isRequired,
  exercise: PropTypes.shape({
    name: PropTypes.string,
    primary: PropTypes.string,
    secondary: PropTypes.string,
  }).isRequired,
  addSetDispatch: PropTypes.func.isRequired,
};

const mapDispatchToProps = (dispatch) => ({
  addSetDispatch: (state, weight, reps, distance, distanceUnit, time, rpe) => {
    const { date, exercise } = state;

    dispatch(addSetAction({
      key: `${date}-${exercise}-${Date.now()}`,
      date,
      exercise: exercise.name,
      weight,
      weightUnit: 'lbs',
      reps,
      distance,
      distanceUnit,
      time,
      rpe,
    }));
  },
});

export default connect(null, mapDispatchToProps)(Footer);
