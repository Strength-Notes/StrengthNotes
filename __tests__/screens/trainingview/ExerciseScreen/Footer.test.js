import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import Footer from '../../../../screens/trainingview/ExerciseScreen/Footer';
import ExerciseProperties from '../../../../redux/ExerciseProperties';

const mockStore = configureMockStore();
const store = mockStore({});

function renderForProperties(primary, secondary) {
  const tree = renderer.create(
    <Provider store={store}>
      <Footer
        date={new Date('2020-08-08')}
        exercise={{
          name: 'Squat',
          primary,
          secondary,
        }}
      />
    </Provider>,
  ).toJSON();
  return tree;
}

describe('<Footer />', () => {
  it('returns the proper component for each ExerciseProperty', () => {
    expect(renderForProperties(
      ExerciseProperties.WEIGHT,
      ExerciseProperties.REPS,
    )).toMatchSnapshot();
    expect(renderForProperties(
      ExerciseProperties.DISTANCE,
      ExerciseProperties.TIME,
    )).toMatchSnapshot();
  });

  it('returns the proper component for undefined properties', () => {
    expect(renderForProperties()).toMatchSnapshot();
  });
});
