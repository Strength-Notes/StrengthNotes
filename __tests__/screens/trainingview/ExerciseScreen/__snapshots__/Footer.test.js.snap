// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Footer /> returns the proper component 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#6f6b70",
      "flexDirection": "row",
      "height": 40,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    keyboardType="decimal-pad"
    maxLength={7}
    onChangeText={[Function]}
    placeholder="Weight"
    rejectResponderTermination={true}
    style={
      Object {
        "backgroundColor": "white",
        "borderRadius": 5,
        "flex": 1,
        "margin": 3,
      }
    }
    underlineColorAndroid="transparent"
  />
  <TextInput
    allowFontScaling={true}
    keyboardType="decimal-pad"
    maxLength={4}
    onChangeText={[Function]}
    placeholder="Reps"
    rejectResponderTermination={true}
    style={
      Object {
        "backgroundColor": "white",
        "borderRadius": 5,
        "flex": 1,
        "margin": 3,
      }
    }
    underlineColorAndroid="transparent"
  />
  <TextInput
    allowFontScaling={true}
    keyboardType="decimal-pad"
    maxLength={4}
    onChangeText={[Function]}
    placeholder="RPE"
    rejectResponderTermination={true}
    style={
      Object {
        "backgroundColor": "white",
        "borderRadius": 5,
        "flex": 1,
        "margin": 3,
      }
    }
    underlineColorAndroid="transparent"
  />
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "margin": 3,
        "marginLeft": 8,
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        Object {
          "backgroundColor": "lightgray",
          "borderRadius": 5,
          "flex": 1,
          "fontWeight": "bold",
          "textAlign": "center",
          "textAlignVertical": "center",
          "width": 81,
        }
      }
    >
      Add
    </Text>
  </View>
</View>
`;

exports[`<Footer /> returns the proper component for each ExerciseProperty 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#6f6b70",
      "flexDirection": "row",
      "height": 40,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    keyboardType="decimal-pad"
    maxLength={7}
    onChangeText={[Function]}
    placeholder="Weight"
    rejectResponderTermination={true}
    style={
      Object {
        "backgroundColor": "white",
        "borderRadius": 5,
        "flex": 1,
        "margin": 3,
      }
    }
    underlineColorAndroid="transparent"
  />
  <TextInput
    allowFontScaling={true}
    keyboardType="decimal-pad"
    maxLength={4}
    onChangeText={[Function]}
    placeholder="Reps"
    rejectResponderTermination={true}
    style={
      Object {
        "backgroundColor": "white",
        "borderRadius": 5,
        "flex": 1,
        "margin": 3,
      }
    }
    underlineColorAndroid="transparent"
  />
  <TextInput
    allowFontScaling={true}
    keyboardType="decimal-pad"
    maxLength={4}
    onChangeText={[Function]}
    placeholder="RPE"
    rejectResponderTermination={true}
    style={
      Object {
        "backgroundColor": "white",
        "borderRadius": 5,
        "flex": 1,
        "margin": 3,
      }
    }
    underlineColorAndroid="transparent"
  />
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "margin": 3,
        "marginLeft": 8,
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        Object {
          "backgroundColor": "lightgray",
          "borderRadius": 5,
          "flex": 1,
          "fontWeight": "bold",
          "textAlign": "center",
          "textAlignVertical": "center",
          "width": 81,
        }
      }
    >
      Add
    </Text>
  </View>
</View>
`;

exports[`<Footer /> returns the proper component for each ExerciseProperty 2`] = `
<View
  style={
    Object {
      "backgroundColor": "#6f6b70",
      "flexDirection": "row",
      "height": 40,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "flex": 1,
          "flexDirection": "row",
        },
        Object {
          "flex": 2,
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      keyboardType="decimal-pad"
      maxLength={7}
      onChangeText={[Function]}
      placeholder="Distance"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": 5,
          "flex": 1,
          "margin": 3,
        }
      }
      underlineColorAndroid="transparent"
    />
    <View
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": 5,
          "flex": 1,
          "margin": 3,
        }
      }
    >
      <RCTPicker
        items={
          Array [
            Object {
              "label": "m",
              "textColor": undefined,
              "value": "m",
            },
            Object {
              "label": "km",
              "textColor": undefined,
              "value": "km",
            },
            Object {
              "label": "yd",
              "textColor": undefined,
              "value": "yd",
            },
            Object {
              "label": "mi",
              "textColor": undefined,
              "value": "mi",
            },
          ]
        }
        onChange={[Function]}
        selectedIndex={0}
        style={
          Array [
            Object {
              "height": 216,
            },
            undefined,
          ]
        }
      />
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "flex": 1,
          "flexDirection": "row",
        },
        Object {
          "flex": 3,
        },
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      keyboardType="decimal-pad"
      maxLength={2}
      onChangeText={[Function]}
      placeholder="HH"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": 5,
          "flex": 1,
          "margin": 3,
        }
      }
      underlineColorAndroid="transparent"
    />
    <TextInput
      allowFontScaling={true}
      keyboardType="decimal-pad"
      maxLength={2}
      onChangeText={[Function]}
      placeholder="MM"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": 5,
          "flex": 1,
          "margin": 3,
        }
      }
      underlineColorAndroid="transparent"
    />
    <TextInput
      allowFontScaling={true}
      keyboardType="decimal-pad"
      maxLength={5}
      onChangeText={[Function]}
      placeholder="SS"
      rejectResponderTermination={true}
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": 5,
          "flex": 1,
          "margin": 3,
        }
      }
      underlineColorAndroid="transparent"
    />
  </View>
  <TextInput
    allowFontScaling={true}
    keyboardType="decimal-pad"
    maxLength={4}
    onChangeText={[Function]}
    placeholder="RPE"
    rejectResponderTermination={true}
    style={
      Object {
        "backgroundColor": "white",
        "borderRadius": 5,
        "flex": 1,
        "margin": 3,
      }
    }
    underlineColorAndroid="transparent"
  />
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "margin": 3,
        "marginLeft": 8,
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        Object {
          "backgroundColor": "lightgray",
          "borderRadius": 5,
          "flex": 1,
          "fontWeight": "bold",
          "textAlign": "center",
          "textAlignVertical": "center",
          "width": 81,
        }
      }
    >
      Add
    </Text>
  </View>
</View>
`;

exports[`<Footer /> returns the proper component for undefined properties 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#6f6b70",
      "flexDirection": "row",
      "height": 40,
    }
  }
>
  <View />
  <View />
  <TextInput
    allowFontScaling={true}
    keyboardType="decimal-pad"
    maxLength={4}
    onChangeText={[Function]}
    placeholder="RPE"
    rejectResponderTermination={true}
    style={
      Object {
        "backgroundColor": "white",
        "borderRadius": 5,
        "flex": 1,
        "margin": 3,
      }
    }
    underlineColorAndroid="transparent"
  />
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "margin": 3,
        "marginLeft": 8,
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        Object {
          "backgroundColor": "lightgray",
          "borderRadius": 5,
          "flex": 1,
          "fontWeight": "bold",
          "textAlign": "center",
          "textAlignVertical": "center",
          "width": 81,
        }
      }
    >
      Add
    </Text>
  </View>
</View>
`;
